### PR TITLE
ARROW-10803: Support R >= 3.3 and add CI

### PR DESCRIPTION
**Pull request by nealrichardson**
Includes fixes for compiling on R 3.4 and R 3.3, plus a crossbow job to build on old R versions

### CODE REVIEW

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@romainfrancois I cherry-picked the relevant commits from #8650 over here

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

File: r/configure [hunk 1]
```diff
@@ -168,6 +162,17 @@ if [ "$ARROW_R_CXXFLAGS" ]; then
 fi
 
 # Test configuration
+CXXCPP="`${R_HOME}/bin/R CMD config CXX11` -E"
```
Comment (jonkeane) [inline r/configure:165]: Would it hurt to have this in the `if [TRUE]` block to match what's in the `else` block?

**Review by jonkeane — COMMENTED** [review]
Looks good, one small note

File: r/configure (hunk 1 above)
Comment (nealrichardson) [reply r/configure:165]: If it succeeded, then it's already set. I'll add a comment.

**Pull request closed without merging**